### PR TITLE
libgpiod based BMDA gpio bitbang backend

### LINKED ISSUE #1063: Direct SWD bit bang programming
Does blackmagic have the capability flash files using SWD pins that are wired directly to an embedded linux processor? As in, there is no USB device. The gpio from the microprocessors SWD pins are connected directly to the main application processor.

### PR DESCRIPTION
**Pull request by TechnoMancer**
## Detailed description

This change adds a new backend/probe type to the "hosted"/BMDA platform which uses libgpiod to connect up the `jtagtap.c` and `swdptap.c` bitbanging routines to gpios.  This is of course linux specific.

* A meson build option is added to enable this backend, this only has an effect on linux and defaults to attempting to find the dependency but disabling the feature if libgpiod is not found.
* A new probe type for gpiod is added in the BMDA code.
* An include file is added which provides mappings from the `gpio_set`, `gpio_clear`, `gpio_get`, `gpio_set_val` routines to the libgpiod backed routines for compiling the jtagtap and swdptap code.
* A new cli option is added when the feature is enabled which allows the user to set a mapping of debug signals to GPIOs if this is specified then it is assumed that the gpio bitbanging backend is to be used.  The format of the mapping is as follows:
    <gpio_mapping>=<signal_mapping>(`,`<signal_mapping>)*
    <signal_mapping>=<debug_signal>`=`<gpio_chip_name>`:`<gpio_offset>
    Where the debug signals are `swdio`, `swclk`, `tms`, `tck`, `tdi`, `tdo` and gpio_chip_name is the name of a gpiochip character device to use and gpio_offset is the numeric offset to the required gpio on that chip.
* The swdptap and jtagtap routines in `/src/platform/common` are built into the BMDA.  This requires conditionally removing the definition of `swd_proc` and `jtag_proc` in the `hosted/platform.c` file as these are usually defined in the relevant tap files.
* The required calls to init and use the relevant routines are added to the BMDA code, this requires the `jtagtap_init` routine to be unconditionally declared by `jtagtap.h`
* The required calls to perform swd or jtag scan are added to the BMDA code.
* This backend sets `target_clk_divider` to `UINT32_MAX` in order to disable delays, the calls to the kernel to change the GPIO status appear to have somewhat significant overhead anyway so the maximum clock rate observed was on the order of kilohertz.
* Similarly no support for getting or setting a debug clock frequency is provided.
* No support for a reset line is included but could be added with a little effort.
* JTAG and SWD lines are assumed disjoint, support for sharing similarly to probe hardware could be implemented later but requires more effort.

The backend has been tested with SWD on a Radxa Rock 4 C+ to a Raspberry Pi Pico.  JTAG has not been tested at time of submitting this PR.

## Your checklist for this pull request

* [x] I've read the [Code of Conduct](https://github.com/blackmagic-debug/blackmagic/blob/main/CODE_OF_CONDUCT.md)
* [x] I've read the [guidelines for contributing](https://github.com/blackmagic-debug/blackmagic/blob/main/CONTRIBUTING.md) to this repository
* [ ] It builds for hardware native (see [Building the firmware](https://github.com/blackmagic-debug/blackmagic?tab=readme-ov-file#building-black-magic-debug-firmware))
* [x] It builds as BMDA (see [Building the BMDA](https://github.com/blackmagic-debug/blackmagic?tab=readme-ov-file#building-black-magic-debug-app))
* [x] I've tested it to the best of my ability
* [x] My commit messages provide a useful short description of what the commits do

## Closing issues

fixes #1063

### CODE REVIEW

File: src/platforms/hosted/bmda_gpiod.h
Comment (dragonmux) [inline src/platforms/hosted/bmda_gpiod.h:42]: The parens here need `void` inside them to avoid declaring an any-parameters function. See style 3 of [function declaration](https://en.cppreference.com/w/c/language/function_declaration) for details.

File: src/platforms/hosted/bmda_gpiod.c
Comment (dragonmux) [inline src/platforms/hosted/bmda_gpiod.c:246]: This and the function below get the same comment about needing `void` in the parens to avoid the potentially bad spicy kind of function definition.

File: src/platforms/hosted/bmda_gpiod.c
Comment (dragonmux) [inline src/platforms/hosted/bmda_gpiod.c:201]: If possible, please initialise this and move it next to the first strtok_r() it gets used in, so we can try and prevent possible future UB read-before-init issues.

File: src/platforms/hosted/bmda_gpiod.c
Comment (dragonmux) [inline src/platforms/hosted/bmda_gpiod.c:140]: Please initialise this to `NULL` as strtoul() may fail and not touch this on some libc's, resulting in UB read-before-init in the if check below.

File: src/platforms/hosted/bmda_gpiod.c
Comment (dragonmux) [inline src/platforms/hosted/bmda_gpiod.c:86]: Please drop the extra braces, they're unnecessary. Same for the instances below for single-statement block bodies.

File: src/platforms/hosted/bmda_gpiod.c
Comment (dragonmux) [inline src/platforms/hosted/bmda_gpiod.c:98]: Using `!!` here is not necessary nor the style of the code base, you may simply let implicit bool conversion do its thing.

**Review by dragonmux — CHANGES_REQUESTED** [review]
This is looking excellent! There are a couple of items to address or at least discuss, and once they're taken care of we'll approve this for merge.

Thank you for the contribution! Hopefully users on RPi's and other similar systems will be glad to be able to use the GPIO facilities built into their SBCs for debug.

File: src/platforms/hosted/bmda_gpiod.c
Comment (TechnoMancer) [reply src/platforms/hosted/bmda_gpiod.c]: I have removed the `!!`s

File: src/platforms/hosted/bmda_gpiod.c
Comment (TechnoMancer) [reply src/platforms/hosted/bmda_gpiod.c]: I think I have sorted this out.

File: src/platforms/hosted/bmda_gpiod.c
Comment (TechnoMancer) [reply src/platforms/hosted/bmda_gpiod.c]: Done

File: src/platforms/hosted/bmda_gpiod.c
Comment (TechnoMancer) [reply src/platforms/hosted/bmda_gpiod.c]: Done

File: src/platforms/hosted/bmda_gpiod.h
Comment (TechnoMancer) [reply src/platforms/hosted/bmda_gpiod.h]: Done

File: src/platforms/hosted/bmda_gpiod_platform.h
Comment (TechnoMancer) [reply src/platforms/hosted/bmda_gpiod_platform.h]: Done

File: src/platforms/hosted/bmda_gpiod.c
Comment (TechnoMancer) [reply src/platforms/hosted/bmda_gpiod.c]: I missed 142 before, thanks.

Should be sorted now.

**Review by dragonmux — APPROVED** [review]
LGTM, merging once the builds complete. Thank you again for the contribution!

**Pull request merged**
